src/index.js

```diff
@@ -19,664 +20,520 @@ const debug = (function () {
+            DISCONNECTED: 'loginFailed',
+            AUTHORIZED: 'authorized',
+            NOT_AUTHORIZED: 'registerNeeded'
+          }
+        },
+        registerNeeded: {
+          invoke: {
+            src: _registerNeeded
+          },
+          on: {
+            MASQ_ACCESS_GRANTED: 'accessMaterialReceived',
+            MASQ_ACCESS_REFUSED: 'loginFailed'
+          }
+        },
+        accessMaterialReceived: {
+          invoke: {
+            src: _registering
+          },
+          on: {
+            WRITE_ACCESS_GRANTED: 'userAppDbCreated'
+          }
+        },
+        userAppDbCreated: {
+          invoke: {
+            src: _dbCreation
+          },
+          on: {
+            CONNECTION_ESTABLISHED: 'connectionEstablished'
+          }
+        },
+        authorized: {
+          invoke: {
+            src: _receivedAuthorized
+          },
+          on: {
+            CONNECTION_ESTABLISHED: 'connectionEstablished',
+            REGISTER_NEEDED: 'registerNeeded'
+          }
+        },
+        loginFailed: {
+          actions: [ _resetLogin ],
+          target: 'notLogged'
+        },
+        connectionEstablished: {
+          invoke: {
+            src: _readSessionInfoIntoState,
+            onDone: 'logged'
+          }
+        },
+        logged: {
+          on: {
+            SIGNOUT: {
+              target: 'notLogged',
+              actions: [_signout]
+            }
+          }
+        }
+      }
+    })
 
-    // init state
-    this.state = 'notLogged'
+    this.service = interpret(this.machine)
+
+    this.service.onTransition(state => {
+      console.log('NEW STATE : ' + state.value)
+    })
 
-    this.initNullAttributes()
+    this.service.start()
+    console.log('STARTED')
 
     // setup listeners
     this._listenForLoginOrSignout()
     this._listenForFocus()
-
-    // init state through async function
-    // this.init should be awaited to make sure state is initialized
-    this.initPromise = this._init()
-  }
-
-  initNullAttributes () {
-    // login state data
-    this.stayConnected = false
-    this.loginSw = null
-    this.loginPeer = null
-
-    // reset replication variables
-    this.replicating = false
-    this.userAppRepSW = null
-    this.userAppRepHub = null
-
-    // reset logged in user variable
-    this.userAppDbId = null
-    this.userAppDb = null
-    this.userAppDEK = null
-    this.importedUserAppDEK = null
-    this.userAppNonce = null
-    this.username = null
-    this.profileImage = null
   }
 
   async init () {
-    await this.initPromise
-  }
-
-  async _init () {
-    debug('[masq._init]')
-    // try to login from session info
-    await this._loadSessionInfo()
-
-    // get login channel and key created at login link generation
-    // sets this.loginKey and this.loginChannel
-    // TODO check login link lifecycle
-    await this._genLoginLink()
+    await new Promise((resolve) => {
+      this.eventTarget.addEventListener('ready', () => {
```

Comment:
Why using 3 types of elements for events listeners ? (here an element, window: https://github.com/QwantResearch/masq-lib/blob/24f1b85deacda386893134d6e546f9b5bb1a94dc/src/index.js#L175), and document https://github.com/QwantResearch/masq-lib/blob/24f1b85deacda386893134d6e546f9b5bb1a94dc/src/index.js#L189

Reply:
ok, I don't know for the focu event :)